Simplify IfThenElse and CompareSelect within for-loop

EikanWang opened the pull request:
Analyze the range to determine if a condition cannot be satisfied. Suppose the for-loop body contains `IfThenElse` or `CompareSelect` while the condition of the two statements depends on the for-loop index `Var`. In that case, we will analyze the range to check whether the condition could always be satisfied or not. If the condition is deterministic, simplify the logic.

EikanWang committed:
Simplify IfThenElse and CompareSelect within for-loop

Analyze the range to determine if a condition cannot be satisfied. Suppose the for-loop body contains `IfThenElse` or `CompareSelect` while the condition of the two statements depends on the for-loop index `Var`. In that case, we will analyze the range to check whether the condition could always be satisfied or not. If the condition is deterministic, simplify the logic.

EikanWang committed:
Simplify IfThenElse and CompareSelect within for-loop

Analyze the range to determine if a condition cannot be satisfied. Suppose the for-loop body contains `IfThenElse` or `CompareSelect` while the condition of the two statements depends on the for-loop index `Var`. In that case, we will analyze the range to check whether the condition could always be satisfied or not. If the condition is deterministic, simplify the logic.

EikanWang committed:
Define BoundInfo for SimplifierUnderContext and use this structure for the condition simplifier

EikanWang committed:
Define BoundInfo for SimplifierUnderContext and use this structure for the condition simplifier

navahgar commented on torch/csrc/jit/tensorexpr/bounds_overlap.h:
```diff
@@ -61,6 +63,12 @@ enum OverlapKind { ContainedOrEqual, Contains, PartialOverlap, NoOverlap };
+//    (a < b) / (a <= b) / (a > b) / (a >= b) / (a == b) / (a != b)
+// Otherwise, it returns false.
+bool TORCH_API
+compareBound(const Bound& a, const Bound& b, BoundCompareResult& cmp_result);
```
Again, we do not use `BoundCompareResult&` to return results from a function. It is something that can be easily overlooked while reading the code. Can you replace this with a pointer instead?

navahgar commented on torch/csrc/jit/tensorexpr/bounds_overlap.cpp:
```diff
@@ -71,6 +71,60 @@ OverlapKind boundOverlap(Bound a, Bound b) {
+    // lhs = [8, 8] and rhs = [5, 10]
+    // lhs = [5, 10] and rhs = [8, 8]
+    // lhs = [5, 10] and rhs = [6, 7]
+    // lhs = [6, 7] and rhs = [5, 10]
```
Shouldn't this example be `kLT`? Similarly, the previous one should be a `kGT`, right?

Are you just being conservative here? If so, can you update the comment here to reflect the reason for that?

navahgar commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -2910,6 +2911,152 @@ ExprPtr SimplifierUnderContext::mutate(DivPtr v) {
+        simplified_ret2,
+        v->compare_select_op(),
+        v->bias());
+  }
```
Why can't we just simplify the entire `CompareSelect` expression instead of doing each of the inner expressions separately?

EikanWang replied:
> Why can't we just simplify the entire `CompareSelect` expression instead of doing each of the inner expressions separately?

Please correct me if I don't understand exactly.
- The `IRSimplifier::simplify` also invokes this mutator https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/tensorexpr/ir_simplifier.cpp#L2968. If we simplify the entire `CompareSelect`, it might lead to infinite recursive invocation.
- Besides that, the `CompareSelect` could be nested, so we need to simplify the inner expression to make sure each `CompareSelect` is simplified.

navahgar replied:
Yes, you are right about that. I missed it earlier. This looks good.

navahgar commented on test/cpp/tensorexpr/test_simplify.cpp:
```diff
@@ -4824,7 +4824,303 @@ TEST(Simplify, SimplifyBroadcastTermExpander) {
+  //   for (const auto n : c10::irange(1, N)) {
+  //     b[n] = 1.f;
+  //   }
+  test_case_fn(n, b, 1, N, 1, kLT, "b[n] = 1.f;");
```
This is actually `n < 1`, which is already tested above. This case, as in comment, should be `n < 0` instead.

EikanWang replied:
Thanks, fixed the UT.

navahgar commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -2910,6 +2911,152 @@ ExprPtr SimplifierUnderContext::mutate(DivPtr v) {
+    // a != b
+    auto ne_true =
+        (cmp_res == analysis::BoundCompareResult::kLT ||
+         cmp_res == analysis::BoundCompareResult::kGT);
```
This has to include `kNE` as well, right?

EikanWang replied:
The `compareBound` does not return `kNE` because it is implied by `kLT` and `kGT`. But I think it makes sense to add `kNE` here to make it easier to read.

navahgar commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -2910,6 +2911,152 @@ ExprPtr SimplifierUnderContext::mutate(DivPtr v) {
+    }
+  };
+
+  auto cmp_op = v->compare_select_op();
```
Can you add some comments to explain the logic here?

navahgar reviewed (commented):
The overall logic looks good to me. I have some comments inlined below.

EikanWang committed:
Polish UT, add more comments and rename BoundInfo to LoopBoundInfo

EikanWang committed:
Polish UT, add more comments and rename BoundInfo to LoopBoundInfo

huiguoo commented on torch/csrc/jit/tensorexpr/bounds_overlap.cpp:
```diff
@@ -71,6 +71,83 @@ OverlapKind boundOverlap(Bound a, Bound b) {
+  if (overlap_kind == OverlapKind::PartialOverlap) {
+    // The element of a and b could not always satisfy the comparison
+    //    a = [2, 10], b = [5, 15]
+    return false;
```
For partial overlaps such as a = [2, 5], b = [5, 10], we still can compare them right? Such cases correspond to kLE or kGE.

EikanWang replied:
> For partial overlaps such as a = [2, 5], b = [5, 10], we still can compare them right? Such cases correspond to kLE or kGE.

Oh, yes. Updated the code.

huiguoo commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -2950,6 +3113,40 @@ ExprPtr SimplifierUnderContext::mutate(ModPtr v) {
 
+bool SimplifierUnderContext::getLoopBoundInfo(
+    const ExprPtr& expr,
+    LoopBoundInfo* loop_bound_info) {
```
For this function, if the bound info can be saved in the form of "Bound" then the definition of "LoopBoundInfo" can be removed. 
```
return analysis::Bound(got->second.first, IRSimplifier::simplify(
      alloc<Sub>(got->second.second, immLike(got->second.second, 1))));
``` 
The definition of "LoopBoundInfo" is so similar to "Bound" in the existing code.

EikanWang replied:
I removed the `LoopBoundInfo` and use `Bound`

huiguoo commented on torch/csrc/jit/tensorexpr/ir_simplifier.h:
```diff
@@ -372,7 +372,9 @@ class MinTerm : public ExprNode<MinTerm> {
 
 // Context-sensitive IR simplification
-using VarBoundInfo = std::unordered_map<VarPtr, std::pair<ExprPtr, ExprPtr>>;
+using LoopBoundInfo = std::pair<ExprPtr, ExprPtr>;
```
It seems that this definition is not necessary. See the comments above.

EikanWang replied:
I removed the `LoopBoundInfo` and use `Bound`

huiguoo commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -2910,6 +2911,168 @@ ExprPtr SimplifierUnderContext::mutate(DivPtr v) {
+  auto _imply = [](const analysis::BoundCompareResult& cmp_res,
+                   const CompareSelectOperation& cmp_op) {
+    // a > b
+    auto gt_true = (cmp_res == analysis::BoundCompareResult::kGT);
+    // a >= b
+    auto ge_true =
+        (cmp_res == analysis::BoundCompareResult::kGE ||
+         cmp_res == analysis::BoundCompareResult::kGT ||
+         cmp_res == analysis::BoundCompareResult::kEQ);
+    // a < b
+    auto lt_true = (cmp_res == analysis::BoundCompareResult::kLT);
+    // a <= b
+    auto le_true =
+        (cmp_res == analysis::BoundCompareResult::kLE ||
+         cmp_res == analysis::BoundCompareResult::kLT ||
+         cmp_res == analysis::BoundCompareResult::kEQ);
+    // a == b
+    auto eq_true = (cmp_res == analysis::BoundCompareResult::kEQ);
+    // a != b
+    auto ne_true =
+        (cmp_res == analysis::BoundCompareResult::kLT ||
+         cmp_res == analysis::BoundCompareResult::kGT ||
+         cmp_res == analysis::BoundCompareResult::kNE);
+
+    switch (cmp_op) {
+      case CompareSelectOperation::kGT:
+        return gt_true;
+      case CompareSelectOperation::kGE:
+        return ge_true;
+      case CompareSelectOperation::kLT:
+        return lt_true;
+      case CompareSelectOperation::kLE:
+        return le_true;
+      case CompareSelectOperation::kNE:
+        return ne_true;
+      default:
+        TORCH_INTERNAL_ASSERT(cmp_op == CompareSelectOperation::kEQ)
+        return eq_true;
+    }
+  };
```
This part is about to compare two bounds with regard to a cmp_op. I feel that it makes more sense to put this logic in the analysis (bounds_overlap.cpp) instead of the simplifier. In the analysis, you can encapsulate this into a function that takes bounds a, bounds b and a cmp_op as inputs, and returns a value that indicates 1) if the cmp_op always holds true for a and b; or 2) if the cmp_op always holds false for a and b; or 3) not determined. 
```
enum CmpOpEvalResult { true, false, notDetermined }
CmpOpEvalResult TORCH_API compareBound(
    const Bound& a,
    const Bound& b,
    CompareSelectOperation& cmp_op);
```  
Feel free to use better names if you decide to do so :) In this way, the logic in simplifier for "CompareSelectPtr" will be much easier, and the analysis can be reused in other places.

EikanWang replied:
I enabled `Bound` to support comparison operators including `<, >, <=, >=, ==, !=`. The comparison is conservative. `True` means all the elements satisfy the given comparison operator. But the `False` does not mean the opposite comparison operator is satisfied. It could be but not always.

huiguoo commented on test/cpp/tensorexpr/test_simplify.cpp:
```diff
@@ -4824,7 +4824,303 @@ TEST(Simplify, SimplifyBroadcastTermExpander) {
 
-TEST(Simplify, DISABLED_CompareSelectCondAlwaysInLoopBounds) {
+TEST(Simplify, CompareSelectLoopBounds) {
+  constexpr int N = 8;
```
@navahgar do we need to test dynamic shapes here? 
```
VarPtr N = alloc<Var>("N", kInt);
```

navahgar replied:
Yes, it will be good to have some tests with dynamic shapes, to ensure that nothing breaks with these changes.

huiguoo commented on test/cpp/tensorexpr/test_simplify.cpp:
```diff
@@ -4824,7 +4824,303 @@ TEST(Simplify, SimplifyBroadcastTermExpander) {
+  test_case_fn(n, b, 1, N, N, kNE, "b[n] = 0.f;");
+}
+
+TEST(Simplify, CompareSelectCondAlwaysInLoopBounds) {
```
For completeness, can you add a test that compares two index variables? For example
```
for (const auto i : c10::irange(1, 5))
  for (const auto j : c10::irange(5, 10))
     b[i, j] = i<=j ? 0.f : 1.f
```

EikanWang replied:
Added more test case.

EikanWang commented on torch/csrc/jit/tensorexpr/ir_simplifier.h:
```diff
@@ -372,7 +372,9 @@ class MinTerm : public ExprNode<MinTerm> {
 
 // Context-sensitive IR simplification
-using VarBoundInfo = std::unordered_map<VarPtr, std::pair<ExprPtr, ExprPtr>>;
+using BoundInfo = std::pair<ExprPtr, ExprPtr>;
```
Per Hui's comments, I reuse the `Bound` structure here.

EikanWang committed:
Simplify the code

EikanWang committed:
Simplify the code

EikanWang committed:
Add more test cases

EikanWang committed:
Add more test cases

huiguoo reviewed (approved):
Thanks for addressing all the comments! It looks great to me now. 

Pls test dynamic shapes and make sure that it won't break anything there :)

EikanWang commented:
> Thanks for addressing all the comments! It looks great to me now.
> 
> Pls test dynamic shapes and make sure that it won't break anything there :)

I added the dynamic shape test cases. Please help review if I need to add more test cases.

EikanWang committed:
Add dynamic shape test cases

EikanWang committed:
Add dynamic shape test cases

huiguoo commented:
@EikanWang Could you rebase the PR to the latest master to see if the 3 CI failures go away? Also let me know if you have problems testing dynamic shapes.

EikanWang commented:
> @EikanWang The dynamic shape test cases look good to me! Thanks for adding them.
> 
> There are 3 CI test failures. I suspect that the failures are not related to the changes made in this PR. To double check on that, could you rebase the PR to the latest master to see if the 3 CI failures go away?

Sure.

EikanWang committed:
Merge branch 'simplify_ifthenelse_compareselect' of https://github.com/EikanWang/pytorch into simplify_ifthenelse_compareselect

EikanWang committed:
Merge branch 'master' into simplify_ifthenelse_compareselect

EikanWang commented:
> > @EikanWang The dynamic shape test cases look good to me! Thanks for adding them.
> > There are 3 CI test failures. I suspect that the failures are not related to the changes made in this PR. To double check on that, could you rebase the PR to the latest master to see if the 3 CI failures go away?
> 
> Sure

Passed all the CI cases.
